[lldb/API] Add setters to SBStructuredData

medismailben opened the pull request:
This patch adds setters to the SBStruturedData class to be able to initialize said object from the client side directly.

jimingham commented on lldb/include/lldb/API/SBStructuredData.h:
```diff
@@ -109,6 +109,22 @@ class SBStructuredData {
   /// Return the generic pointer if this data structure is a generic type.
   lldb::SBScriptObject GetGenericValue() const;
 
+  void SetValueForKey(const char *key, SBStructuredData &value) const;
```
Also, while it's kind of cool that you are able to fool the compiler into thinking these methods are const, the very much change the object in question, so I'm not sure marking them as const is helpful.

jimingham replied:
I wasn't suggesting that.  I was suggesting doing what you do for all the other "Set" API.  If an SBStructuredData was actually a dictionary, and someone calls "SetFloatValue" on it, you don't attempt to merge the new float value with the extant dictionary; instead you discard the dictionary and make the SBStructuredData hold a Float type.  Similarly, if I had an SBStructuredData that was an Integer type, and you said "SetValueForKey", I was suggesting you discard the Integer object and make it hold a dictionary with one element, the one passed in to SetValueForKey.

bulbazord replied:
If I made a mistake and accidentally called `SetValueForKey` on the wrong `SBStructuredData`, it would destroy what I already have. That would definitely be a frustrating bug.

Might I suggest adding an `SBError` return value? If you try to set an integer value on a dictionary, it would return an error that you've tried an incompatible operation.

To monologue further, I also recognize that this is a huge difference between C++ and python. Jim's suggestion would make a lot of sense for Python where you're slinging data around and types only matter when you crash. In C++, I think error-checking is often baked into APIs in one way or another. I personally prefer when my type system helps me catch issues statically, but I can understand if you'd prefer to keep your proposed behavior.

jimingham replied:
I don't have a strong opinion which way to go.  I feel more strongly that we should be consistent and always stomp the contents and replace them with whatever you asked for or never do.

None of this is going to crash in C++ because all the interfaces to the StructuredData or SBStructuredData check the types behind the scenes.  If you accidentally clobbered a StructuredData that had an integer value with a Dictionary and then asked for the Integer value, you'd just get told it wasn't an Integer.  So this is not crashy and the type system itself wouldn't really help you either.  But I agree it is error prone.

 If you go the way Alex is suggesting and have these API's never convert the underlying object type, then you should also return an error rather than void.

jimingham commented on lldb/include/lldb/API/SBStructuredData.h:
```diff
@@ -109,6 +109,35 @@ class SBStructuredData {
 
+  /// Set the value corresponding to a key. If this data structure
+  /// is not a dictionary type, reset the type to be dictionary and overwrite
+  /// previous data.
```
"the previous data"

jimingham commented:
Other than that this LGTM.

medismailben committed:
[lldb/API] Add setters to SBStructuredData

This patch adds setters to the SBStruturedData class to be able to
initialize said object from the client side directly.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>
